pkg/container-utils/containerd/containerd.go

```diff
@@ -15,31 +15,213 @@
 package containerd
 
 import (
+	"context"
+	"fmt"
 	"time"
 
-	criclient "github.com/inspektor-gadget/inspektor-gadget/pkg/container-utils/cri"
+	"github.com/containerd/containerd"
+	tasks "github.com/containerd/containerd/api/services/tasks/v1"
+	"github.com/containerd/containerd/api/types/task"
+
 	runtimeclient "github.com/inspektor-gadget/inspektor-gadget/pkg/container-utils/runtime-client"
 )
 
 const (
 	DefaultTimeout = 2 * time.Second
+
+	LabelK8sContainerName         = "io.kubernetes.container.name"
+	LabelK8sContainerdKind        = "io.cri-containerd.kind"
+	LabelK8sContainerdKindSandbox = "sandbox"
 )
 
 type ContainerdClient struct {
-	criclient.CRIClient
+	client *containerd.Client
 }
 
 func NewContainerdClient(socketPath string) (runtimeclient.ContainerRuntimeClient, error) {
 	if socketPath == "" {
 		socketPath = runtimeclient.ContainerdDefaultSocketPath
 	}
 
-	criClient, err := criclient.NewCRIClient(runtimeclient.ContainerdName, socketPath, DefaultTimeout)
+	client, err := containerd.New(socketPath,
+		containerd.WithTimeout(DefaultTimeout),
+		containerd.WithDefaultNamespace("k8s.io"),
+	)
 	if err != nil {
 		return nil, err
 	}
 
 	return &ContainerdClient{
-		CRIClient: criClient,
+		client: client,
+	}, nil
+}
+
+func (c *ContainerdClient) Close() error {
+	if c.client != nil {
+		return c.client.Close()
+	}
+	return nil
+}
+
+func (c *ContainerdClient) isSandboxContainer(container containerd.Container) bool {
+	labels, err := container.Labels(context.TODO())
+	if err != nil {
+		return false
+	}
+
+	if kind, ok := labels[LabelK8sContainerdKind]; ok {
+		return kind == LabelK8sContainerdKindSandbox
+	}
+
+	return false
+}
+
+func (c *ContainerdClient) GetContainers() ([]*runtimeclient.ContainerData, error) {
+	taskResponse, err := c.client.TaskService().List(context.TODO(), &tasks.ListTasksRequest{})
+	if err != nil {
+		return nil, err
+	}
+
+	ret := make([]*runtimeclient.ContainerData, 0, len(taskResponse.Tasks))
+	for _, task := range taskResponse.Tasks {
+
+		container, err := c.getContainerdContainer(task.ID)
+		if err != nil {
+			return nil, err
+		}
+
+		if c.isSandboxContainer(container) {
+			continue
+		}
+
+		containerData, err := c.containerdTaskAndContainerToContainerData(task, container)
+		if err != nil {
+			return nil, err
+		}
+
+		ret = append(ret, containerData)
+	}
+
+	return ret, nil
+}
+
+func (c *ContainerdClient) GetContainer(containerID string) (*runtimeclient.ContainerData, error) {
+	response, err := c.client.TaskService().Get(context.TODO(), &tasks.GetRequest{
+		ContainerID: containerID,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	containerData, err := c.ContainerdTaskToContainerData(response.Process)
```

Comment:
Hmmm, do we want verification here? What if its a sandbox container? Should we `return nil, nil` ?

Reply:
Yep, we need to do that verification here. Otherwise, the containers created at runtime will be added without enrichment unless another enricher enriches it or drops it, which is precisely what is happening here. 

After adding a few logs https://github.com/inspektor-gadget/inspektor-gadget/commit/fc5418fe6e2c31f8bfc73a649ddddc981373aa7c, I realized that it was the `OCIConfigEnricher` who was dropping the sandbox containers and enriching the normal containers (because the containerd enricher was failing). My test:

```bash
$ sudo ./local-gadget list-containers -w -v -r containerd
[...]
# Already existing containers
[...]
# Create a pod in another terminal, e.g. kubectl run --rm --image=busybox mypod -it --restart='Never' -- sh
DEBU[0003] runc fanotify: add container "27a08e8e49d1f222b6aa27fa2f47f8418439a2da91cb7bf860f859ee7c18fb5f"
DEBU[0003] OCIConfig enricher: temporary dropping sandbox container "27a08e8e49d1f222b6aa27fa2f47f8418439a2da91cb7bf860f859ee7c18fb5f"
DEBU[0004] runc fanotify: add container "9dd55e85a2596eec5f852cdb616c2b48618a67997d469ac6c6f71ef36b4bf4e9"
DEBU[0004] OCIConfig enricher: container "9dd55e85a2596eec5f852cdb616c2b48618a67997d469ac6c6f71ef36b4bf4e9" was enriched
DEBU[0004] Runtime enricher (containerd): failed to get container: listing task for container "9dd55e85a2596eec5f852cdb616c2b48618a67997d469ac6c6f71ef36b4bf4e9": rpc error: code = NotFound desc = task 9dd55e85a2596eec5f852cdb616c2b48618a67997d469ac6c6f71ef36b4bf4e9 not found
2023-01-10T09:31:49Z           CREATED    containerd 9dd55e85a2596eec5f852cdb616c2b48618a67997d469ac6c6f71ef… mypod
```

Notice that if you remove the [OCIConfigEnricher](https://github.com/inspektor-gadget/inspektor-gadget/blob/98649ab9ea69ff9b7915347882dcd88385fa267d/pkg/local-gadget-manager/local-gadget-manager.go#L351), the containerd enricher will add the sandbox and the normal container without enrichments:

```bash
$ sudo ./local-gadget list-containers -w -v -r containerd -o json | jq .container.podname
[...]
# Already existing containers
"kube-controller-manager"
"coredns"
"kube-proxy"
[...]
# Create a pod in another terminal, e.g. kubectl run --rm --image=busybox mypod -it --restart='Never' -- sh
DEBU[0007] runc fanotify: add container "7761902bd3ff3e333fa3a2c0f443bd5e75debfa9fdd1d3f8c8c5cad42a868849"
DEBU[0007] Runtime enricher (containerd): failed to get container: listing task for container "7761902bd3ff3e333fa3a2c0f443bd5e75debfa9fdd1d3f8c8c5cad42a868849": rpc error: code = NotFound desc = task 7761902bd3ff3e333fa3a2c0f443bd5e75debfa9fdd1d3f8c8c5cad42a868849 not found
null
DEBU[0008] runc fanotify: add container "14e6f198d600361dbd9ffb9f451a65ff9be15e1d41670dea34971eb3fdb0590f"
DEBU[0008] Runtime enricher (containerd): failed to get container: listing task for container "14e6f198d600361dbd9ffb9f451a65ff9be15e1d41670dea34971eb3fdb0590f": rpc error: code = NotFound desc = task 14e6f198d600361dbd9ffb9f451a65ff9be15e1d41670dea34971eb3fdb0590f not found
null
```

This is why I said we should add some tests to verify that each enricher is working correctly. It is because in the previous test, the `OCIConfigEnricher` was hiding the bug in the container enricher.

To summarize, the previous tests show two issues:

1. Containerd enricher should drop sandbox containers also in `GetContainer`. You could use the same strategy used in [Docker](https://github.com/inspektor-gadget/inspektor-gadget/blob/98649ab9ea69ff9b7915347882dcd88385fa267d/pkg/container-utils/docker/docker.go#L95-L97), or directly solve the [main issue](https://github.com/inspektor-gadget/inspektor-gadget/issues/1095) to avoid adding another workaround.
2. When `GetContainer` is called, the container task hasn't been created yet, so the containerd enricher fails to enrich the containers created at runtime. The task is only needed to get the container status, and for `GetContainer`, we don't need the status as we already know that the container is being created. Therefore, in `GetContainer` we can avoid getting the task and we can get the rest of the information with `client.Containers(context.TODO(), fmt.Sprintf("id==%s", id))`. WDYT?

Reply:
Thanks for the 

1. I think that it should be done in another PR. Will do that later, will add this temp workaround.
2. I remember that `nerdctl` did something. The [comment in this line](https://github.com/containerd/nerdctl/blob/b0a75d880ef9e6d1f1a8752804c9087ee2d02f73/pkg/formatter/formatter.go#L48) rightfully mentions that the status should be at least "created" instead of failing to enrich. I'll change the code to handle that gracefully